src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1441,6 +1448,29 @@ public Solution WithAnalyzerConfigDocumentTextLoader(DocumentId documentId, Text
             return newCompilationState == _compilationState ? this : new Solution(newCompilationState);
         }
 
+        /// <summary>
+        /// Returns a solution instance where every project is frozen at whatever current state it is in
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        internal Task<Solution> WithFrozenPartialCompilationsAsync(CancellationToken cancellationToken)
+            => _cachedFrozenSolution.GetValueAsync(cancellationToken);
+
+        private async Task<Solution> ComputeFrozenSolutionAsync(CancellationToken cancellationToken)
+        {
+            // in progress solutions are disabled for some testing
+            if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
+                return this;
+
+            var newCompilationState = await this.CompilationState.WithFrozenPartialCompilationsAsync(cancellationToken).ConfigureAwait(false);
+            var frozenSolution = new Solution(newCompilationState);
+
+            // Set the frozen solution to be its own frozen solution.  Freezing multiple times is a no-op.
+            frozenSolution._cachedFrozenSolution = _cachedFrozenSolution;
```

Comment:
Do you want to add a different constructor so this field doesn't have to be writable? Or can we make it init-only or something?

Reply:
sure.  i can do that.